v0.28: API key updates (#1738)
meilisearch/documentation: reference/api/keys.md

# Keys

> The `/keys` route allows you to create, manage, and delete API keys. To use these endpoints, you must first [set the master key](/learn/configuration/instance_options.md#master-key). Once a master key is set, you can access these endpoints by supplying it in the header of the request, or use API keys that have access to the `keys.get`, `keys.create`, `keys.update`, or `keys.delete` actions.

Author: ```suggestion
The `/keys` route allows you to create, manage, and delete API keys. To use these endpoints, you must first [set the master key](/learn/configuration/instance_options.md#master-key). Once a master key is set, you can access these endpoints by supplying it in the header of the request, or using API keys that have access to the `keys.get`, `keys.create`, `keys.update`, or `keys.delete` actions.
```

> The `/keys` route allows you to create, manage, and delete API keys. To use these endpoints, you must first [set the master key](/learn/configuration/instance_options.md#master-key).Once a master key is set, you caan access these endpoints by and supplying it in the header of the request, or use API keys that have access to the `keys.get`, `keys.create`, `keys.update`, or `keys.delete` actions.

Author: ```suggestion
The `/keys` route allows you to create, manage, and delete API keys. To use these endpoints, you must first [set the master key](/learn/configuration/instance_options.md#master-key). Once a master key is set, you can access these endpoints by supplying it in the header of the request, or use API keys that have access to the `keys.get`, `keys.create`, `keys.update`, or `keys.delete` actions.
```

[Learn more about managing keys and their rights](/learn/security/master_api_keys.md).

## Get all keys

<RouteHighlighter method="GET" route="/keys"/>

> List all existing API keys. **Expired keys are included in the response**, but deleted keys are not. Results can be paginated by using the `offset` and `limit` query parameters.

Reviewer: We may want to mention that the default `limit` is `20` by default here.

Author: Good catch, thanks!

> List the most recently created 20 keys. **Expired keys are included in the response**, but deleted keys are not. Results can be paginated by using the `offset` and `limit` query parameters.

Reviewer 2: Same as above. Perhaps it's just a matter of preference?

Googling "the 20 most beautiful" gives me 290.000 results, whereas "the most beautiful 20" gives me 28.000 🤔  (I tried using a sentence a bit more common than "recently created keys").

```suggestion
List the 20 most recently created keys. **Expired keys are included in the response**, but deleted keys are not. Results can be paginated by using the `offset` and `limit` query parameters.
```

      "updatedAt": "2021-10-13T15:00:00Z"
    },
    {
      "name": "Default Search API Key",

>       "description": "Default Search API Key (Use it to search from the frontend code)",

Reviewer: ```suggestion
      "description": "Use it to search from the frontend code",
```

      "updatedAt": "2021-08-11T10:00:00Z"
    },
    {
      "name": "Default Admin API Key",

>       "description": "Default Admin API Key (Use it for all other operations. Caution! Do not share it on the client side)",

Reviewer: ```suggestion
      "description": "Use it for anything that is not a search operation. Caution! Do not expose it on a public frontend",
```

      "createdAt": "2021-08-11T10:00:00Z",
      "updatedAt": "2021-08-11T10:00:00Z"
    }
  ],
  "offset":0,
  "limit":3,
  "total":7
}
```

::: note
API keys are displayed in descending order based on their `createdAt` date. This means that the most recently created keys appear first.
:::

### Returned fields

Returns API keys in an array called `results`, along with the following fields:

#### `offset`

The number of keys skipped over.

#### `limit`

The maximum number of keys to be returned by the request.

#### `total`

The total number of API keys that can be browsed.

### The `results` array

For each key, it returns:

#### `name`

A human-readable name for the key. Default value is `null`.

#### `description`

A description for the key. Default value is `null`.

#### `uid`

> A [uuid v4](https://www.sohamkamani.com/uuid-versions-explained) to identify the API key. If not specified, it is generated by Meilisearch.

Reviewer 2: I feel it's slightly more clear if we specify it's something that happens automatically.

```suggestion
A [uuid v4](https://www.sohamkamani.com/uuid-versions-explained) to identify the API key. If not specified, it is automatically generated by Meilisearch.
```

      "createdAt": "2021-08-11T10:00:00Z",
      "updatedAt": "2021-08-11T10:00:00Z"
    }
  ],
  "offset":0,
  "limit":3,
  "total":7
}
```

::: note
API keys are displayed in descending order based on their `createdAt` date. This means that the most recently created keys appear first.
:::

### Returned fields

Returns API keys in an array called `results`, along with the following fields:

#### `offset`

The number of keys skipped over.

#### `limit`

The maximum number of keys to be returned by the request.

#### `total`

The total number of API keys that can be browsed.

### The `results` array

For each key, it returns:

#### `name`

A human-readable name for the key. Default value is `null`.

#### `description`

A description for the key. Default value is `null`.

#### `uid`

A [uuid v4](https://www.sohamkamani.com/uuid-versions-explained) to identify the API key. If not specified, it is automatically generated by Meilisearch.

#### `key`

> An alphanumeric key value generated by Meilisearch by hasing the `uid` and the master key on API key creation. Used for authorization when [making calls to a protected Meilisearch instance](/learn/security/master_api_keys.md#communicating-with-a-protected-instance).

Reviewer 3: ```suggestion
An alphanumeric key value generated by Meilisearch by hashing the `uid` and the master key on API key creation. Used for authorization when [making calls to a protected Meilisearch instance](/learn/security/master_api_keys.md#communicating-with-a-protected-instance).
```

This value is also used as the `{key}` path variable to [update](#update-a-key), [delete](#delete-a-key), or [get](#get-one-key) a specific key.

::: note
Since `key` is a hash of the `uid` and master key, `key` values are deterministic between instances sharing the same configuration. This means if the master key changes, all key values are automatically changed and that API key resources are propagated into dumps and snapshots without displaying the `key` field in clear.

> :::

Reviewer 3: Nice explanation here ✨ 
If a snapshot or a dump leaks into nature, nobody would be able to know the key value without knowing the master key. 👍

Reviewer: Regarding this part @meilisearch/docs-team, it could be super cool to give users a bash command to know the `key` field regarding the master-key value and the uid. 

It could be useful for continuous deployment. See https://github.com/meilisearch/product/discussions/421#discussioncomment-3079410

The core team is looking at this to see what can be proposed!

Author: @gmourier, I added the bash command [here](https://github.com/meilisearch/documentation/pull/1738/files#diff-d40e36ee90d9727237e72205b40af7ea904d5d0ce0b277774e7c5ebe8501bdf3). Could you please review this part? 🙏

> Since `key` is a hash of the `uid` and master key, `key` values are deterministic between instances sharing the same configuration. This means if the master key changes, all `key` values are automatically changed. The `key` field is computed at runtime and therefore is not propagated to dumps and snapshots. Even if someone does have access to your dumps or snapshot, they cannot make malicious requests.

Reviewer 2: What do you think about this?

```suggestion
Since `key` is a hash of the `uid` and master key, `key` values are deterministic between instances sharing the same configuration. This means if the master key changes, all `key` values are automatically changed. 

Since the `key` field is depends on the master key, it is computed at runtime and therefore not propagated to dumps and snapshots. As a result, even if a malicious user comes into possession of your dumps or snapshots, they will not have access to your instance's API keys.
```

Author: Thanks Gui!

>     "description": "Add documents: Products API key",

Reviewer 2: Shouldn't we have the `name` and `uid` fields in this response?

## Update a key

> <RouteHighlighter method="PATCH" route="/keys/{uid}"/>

Reviewer: Good question. There is no particular reason, it has not been thought out. We only imagined it for GET.

It would indeed be interesting to facilitate the user experience. WDYT @Kerollmops is it a change that can be made?

> <RouteHighlighter method="PATCH" route="/keys/{uid}"/>

Reviewer 4: Technically speaking it must be possible to be able to update a key by either specifying the UID or the key itself. I don't know why we would allow both to be specified thought.

> <RouteHighlighter method="PATCH" route="/keys/{uid}"/>

Reviewer: You know what? The core team managed it directly. @curquiza and I just checked it out. I will update the spec! 

It is possible to specify an API Key resource for the PATCH and DELETE using `:uid_or_key` as well on the GET verb.

We can find some happiness in misfortune. 😄

> <RouteHighlighter method="PATCH" route="/keys/{uid}"/>

Reviewer 2: Consistency makes me happy. We'll update the docs accordingly.

```json
{
    "name": "Products/Reviews API key",
    "description": "Manage documents: Products/Reviews API key",
    "key": "d0552b41536279a0ad88bd595327b96f01176a60c2243e906c52ac02375f9bc4",
    "uid": "6062abda-a5aa-4414-ac91-ecd7944c0f8d",
    "actions": [
      "documents.add",
      "documents.delete"
    ],
    "indexes": [
      "products",
      "reviews"
    ],
    "expiresAt": "2021-12-31T23:59:59Z",
    "createdAt": "2021-10-12T00:00:00Z",

>     "updatedAt": "2021-10-13T15:00:00Z"

Author: I used the key from line 35 so I updated all the values
